### PR TITLE
Use tables to summarize course content

### LINKED ISSUE #1982: Timing information enters the translations
Hi @djmitche, I noticed in #1974 that the timing information enters the translations. This is in some sense working as intended, but I fear that it creates extra work for the translators.

On [Welcome to Day 1](https://google.github.io/comprehensive-rust/welcome-day-1.html#welcome-to-day-1), as an example, if we would format

- [Welcome](https://google.github.io/comprehensive-rust/welcome-day-1.html) (5 minutes)
- [Hello, World](https://google.github.io/comprehensive-rust/hello-world.html) (15 minutes)
- [Types and Values](https://google.github.io/comprehensive-rust/types-and-values.html) (40 minutes)
- [Control Flow Basics](https://google.github.io/comprehensive-rust/control-flow-basics.html) (40 minutes)

as a table

| Page | |
| --- | ---:|
| Welcome | 5 minutes |
| Hello, World | 15 minutes |
| Types and Values | 40 minutes |
| Control Flow Basics | 40 minutes |

then the page titles would be extracted on their own — and they would match the title we already extract from each page. So it would reuse the translation for the page, which is kind of neat :slightly_smiling_face: 

I don't have a good idea for how to include the page link like in the current formatting. We include the link in the translatable message, so this would defeat the attempt at unifying the message here with the page title.

### PR DESCRIPTION
**Pull request by djmitche**
This is more friendly to translation (as it can share the translation of the title).

This fixes #1982.

### CODE REVIEW

File: mdbook-course/src/course.rs [hunk 1]
```diff
@@ -394,28 +382,19 @@ impl Segment {
         self.into_iter().map(|s| s.minutes()).sum()
     }
 
-    pub fn outline(&self, at_source_path: impl AsRef<Path>) -> String {
-        let mut outline = String::from("In this segment:\n");
+    pub fn outline(&self) -> String {
+        let mut slides = Table::new(["Slide".into(), "Time".into()]);
```
Comment (mgeisler) [inline mdbook-course/src/course.rs:386]: Would it be slightly more accurate to call this a "duration"?

```suggestion
        let mut slides = Table::new(["Slide".into(), "Duration".into()]);
```

File: mdbook-course/src/course.rs [hunk 2]
```diff
@@ -313,24 +307,18 @@ impl Session {
...
-                duration(segment.minutes())
-            )
-            .unwrap();
+            segments.add_row([segment.name.clone(), duration(segment.minutes())]);
         }
-        writeln!(&mut outline,"\nIncluding {BREAK_DURATION} minute breaks, this session should take about {}", duration(self.minutes())).unwrap();
-        outline
+        format!(
+            "Including {BREAK_DURATION} minute breaks, this session should take about {}. It contains:\n\n{}",
+            duration(self.minutes()),segments)
```
Comment (mgeisler) [inline mdbook-course/src/course.rs:321]: I think your `rustfmt` has given up here :smile: 

```suggestion
            duration(self.minutes()), segments)
```

File: mdbook-course/src/course.rs [hunk 3]
```diff
@@ -313,24 +307,18 @@ impl Session {
 
     /// Generate a Markdown outline for this session, for placement at the given
     /// path.
-    pub fn outline(&self, at_source_path: impl AsRef<Path>) -> String {
-        let mut outline = String::from("In this session:\n");
+    pub fn outline(&self) -> String {
+        let mut segments = Table::new(["Segment".into(), "Time".into()]);
```
Comment (mgeisler) [inline mdbook-course/src/course.rs:311]: Perhaps "duration" is better here?

```suggestion
        let mut segments = Table::new(["Segment".into(), "Duration".into()]);
```

File: mdbook-course/src/course.rs [hunk 4]
```diff
@@ -245,32 +245,26 @@ impl Course {
...
         for session in self {
             writeln!(
                 &mut outline,
-                " * {} ({}, including breaks)",
+                " * {} ({}, including breaks)\n",
                 session.name,
                 duration(session.minutes())
             )
             .unwrap();
+            let mut segments = Table::new(["Segment".into(), "Time".into()]);
```
Comment (mgeisler) [inline mdbook-course/src/course.rs:258]:
```suggestion
            let mut segments = Table::new(["Segment".into(), "Duration".into()]);
```

File: mdbook-course/src/markdown.rs [hunk 5]
```diff
@@ -57,6 +58,46 @@ pub fn duration(mut minutes: u64) -> String {
     }
 }
 
+/// Table implements Display to format a two-dimensional table as markdown,
+/// following https://github.github.com/gfm/#tables-extension-.
+pub struct Table<const N: usize> {
```
Comment (mgeisler) [inline mdbook-course/src/markdown.rs:63]: Nice! If you have time, it would be nice to have a small unit test below to show how the output looks like.

**Review by mgeisler — APPROVED** [review]
Thanks @djmitche!

File: mdbook-course/src/course.rs (hunk 2 above)
Comment (djmitche) [reply mdbook-course/src/course.rs:321]: I think it doesn't do much in macro invocations? I was surprised by this too.

**Pull request merged**